Elevation calculation speed-up

evansiroky opened the pull request:
To be completed by pull request submitter:

- [x] **issue**: #2583 seems relevant
- [x] **roadmap**: Not on roadmap :(
- [x] **tests**: Test added for elevation module
- [x] **formatting**:
- [x] **documentation**: If you are adding a new configuration option, have you added an explanation to the [configuration documentation](docs/Configuration.md) tables and sections?
- [x] **changelog**: add a bullet point to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Changelog.md) with description and link to the linked issue

To be completed by @opentripplanner/plc:

- [ ] reviews and approvals by 2 members, ideally from different organizations
- [ ] **after merging**: update the relevant card on the [roadmap](https://github.com/orgs/opentripplanner/projects/1)

This PR dramatically speeds up the calculations in the ElevationModule of the graph builder modules. It does so by:

- Optimizing as many calculations of elevations for a street edge as possible.
- Adding the ability to read/write cached elevation data
  - After graph building, a file with a lookup of the cached elevation for a coordinate sequence can be written
  - In subsequent graph builds, this file can be loaded and the cached data can be used.
  - custom graph build parameters can be entered to enable or disable reading or writing of this data. The default is to read, but not to write.

In addition to the performance improvements, this PR also has changes that:
- tolerate areas where elevation tiles are unavailable.
- refactor a few elevation classes to simplify constructors
- Add a simple test of the elevation module
- adds documentation of how to optimize elevation calculations

t2gran commented:
@evansiroky I ignored the Bano Geocoder test in the `dev-1.x, so you should be able to merge that into your branch.

evansiroky commented:
Thanks, @t2gran. I just did that.

gmellemstrand commented:
I have tested the parallelization aspect of this PR, but not the caching. I tried building a graph with the Oslo OSM data and the Norwegian elevation data. I tried building without the graph on the dev-1.x branch and on this PR branch.

I could not see any speed up in graph build time, and when profiling it seems that many of the threads created by the parallell stream are blocked. Could there be too much synchronization both in the foreach itself and in processEdge method?

evansiroky commented:
@gmellemstrand, are you able to share the osm and elevation data you are using so I can try locally?

evansiroky commented:
I also just pushed a change that should remove some thread blocking.

gmellemstrand commented:
@evansiroky 

Norway height data: https://storage.googleapis.com/marduk/kartverket/heighDataOslo/6602_50m_33.tif
Oslo OSM data: https://storage.googleapis.com/marduk/osm/oslo_norway.osm-160816.pbf

Building with dev-1.x: 8.7 minutes.
Bulding with dev-1x-elevation-improvements: 11.4 minutes.

Profiling with YourKit: https://drive.google.com/open?id=16s3CMbF_qhng8bYP2_HGKMGNqnjwYCh3 (the red blocks show waiting for synchronization)

I also tested with removing various lines that could require synchronization (also within the processEdge method), but I was not able to find what was causing it. It seems like it _should_ be able to work, as processing the edges should be independent of eachother.

evansiroky commented:
Thanks. I am seeing what you are seeing and believe I have identified the source of all the thread locking. I will implement a fix today.

evansiroky commented:
I just pushed a few commits that definitely parallelize the elevation calculations for the Norway elevation config. However, in my own testing, I noticed some deadlocks occurring, so I hope I have implemented some additional code that prevents that from happening.

I have also just added the ability to precompute the (approximate) EllipsoidToGeoidDifference for each elevation point in the graph. Documentation about setting that up coming tomorrow.

gmellemstrand commented:
I just did a test and I don't see any synchronization issues anymore. Total graph build time was 5.9 minutes, but I expect the difference to be much bigger when building the entire Norway graph, because in that case most of the time is spent processing elevation data on the edges.

optionsome commented:
Is this ready for review again? I could take a look at this at some point this week.

evansiroky commented:
Yes, it is ready for review. I kept discovering more opportunities for improvement, but I think it's at a good point now.

evansiroky commented:
Implemented one more improvement, still ready for review.

t2gran commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -80,55 +129,175 @@ public ElevationModule(ElevationGridCoverageFactory factory, double elevationUni
+                coveragesForThread.put(currentThreadId, threadSpecificCoverage);
+            }
+            return threadSpecificCoverage;
+        }
```
It looks like the `coveragesForThread` is a map for each Thread to store some "data", another way to achieve it would be to create a `ForkJoinWorkerThreadFactory` which return a special thread with this data set. Then the map is absolute and the data only visible to the "correct" thread.

evansiroky commented on src/main/java/org/opentripplanner/graph_builder/module/ned/DegreeGridNEDTileSource.java:
```diff
@@ -120,18 +119,10 @@ private File getPathToTile(int x, int y) {
+                log.error("Error downloading tile {}! Error: {}.", key, e.getMessage());
                 path.deleteOnExit();
-                throw new RuntimeException(e);
+                return null;
```
The reason this was done here was that we simply didn't have the needed elevation tiles at certain edges of our graph. Therefore, we needed OTP to be able to continue when a certain tile was requested that we didn't have available.

t2gran replied:
That is understandable - should this be a build annotation, instead then?

abyrd replied:
I think it's acceptable to continue after an exception _if_ we distinguish between network failures / server errors (where we don't know whether the tile exists) and successful communication indicating there is no tile. So I think we should keep separate catch clauses, and fail hard or retry on exceptions representing communication/server failures, while soft-failing (continuing) on exceptions representing successful communication that a file is not present on the remote server.

evansiroky replied:
That seems like a good idea.

evansiroky commented on docs/Configuration.md:
```diff
@@ -230,6 +232,29 @@ it is possible to define a multiplier that converts the elevation values from so
+In graph builds, the elevation module will attempt to read the `cached_elevations.obj` file from the cache directory. The cache directory defaults to `/var/otp/cache`, but this can be overriden via the CLI argument `--cache <directory>`. For the same graph build for multiple Northeast US states, the time it took with using this predownloaded and precalculated data became 543.7 seconds (roughly 9  ...
+
+The cached data is a lookup table where the coordinate sequences of respective street edges are used as keys for calculated data. Therefore, it is expected that over time various edits to OpenStreetMap will cause this cached data to become stale and not include new OSM ways. Therefore, periodic update of this cached data is recommended.
+
```
I don't understand what you mean by `accessed` flag. The elevation cache is only read from. If the `writeCachedElevations` build parameter is set, then a new map is created and written to a file.

t2gran replied:
Ok, that would have the same effect. Keep it like it is, then.

evansiroky commented:
@t2gran, I have refactored using the ForkJoinWorkerThreadFactory and implemented a few other items as you suggested. Also, see my comments for other items I have questions on.

t2gran commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -222,22 +231,69 @@ public void buildGraph(Graph graph, HashMap<Class<?>, Object> extra) {
+                        new DirectPosition2D(GeometryUtils.WGS84_XY, firstEdgeCoord.x, firstEdgeCoord.y),
+                        dummy
+                    );
+                }
```
Do the call to the `evaluate` method need to be called with the first coordinate from the `swee` or just any coordinate? If any coordinate will do it, then you could create the `Coverage` in the constructor and make the  `threadSpecificCoverage` `final`. Also, I think the `synchronized` can be dropped in that case, because the creation of a new Thread should be thread-safe.

_Note! The above code is NOT thread-safe you check the `threadSpecificCoverage == null` outside of the synchronized block and assign it inside._

evansiroky replied:
I have pushed a change that will hopefully resolve the issues you brought up. The reasoning behind not immediately creating a coverage instance is that it can cause a lot of blocking in the beginning when the ForkJoinPool begins. When using the cached elevations, it is highly likely that most edges will already have a value present in the cache and therefore not even need a coverage instance. By lazy-creating the coverage instances other threads can continue to process edges that are available in the cache that don't need a coverage instance, otherwise, a lot of threads will be waiting for each other at the start until a coverage instance has been calculated for each thread.

Regarding the thread-safe assignment, I don't fully understand how it would not be thread-safe to assign the value in this way since the field would only be accessed by the thread doing the assignment and not by anything else. Nonetheless, I attempted to break out the assignment of the variable as suggested, so hopefully, this now makes it thread-safe.

t2gran replied:
Sorry, I forgot the `threadSpecificCoverage` was inside the thread. It looks ok then. You could also have achieved this with a `ThreadLocal` but this good. The main point is to avoid every task to wait on the `synchronized (gridCoverageFactory)`.

t2gran commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -161,7 +169,7 @@ public void buildGraph(Graph graph, HashMap<Class<?>, Object> extra) {
+            parallelism,
             forkJoinWorkerThreadFactory,
             null,
             false
```
Do you know if creating a ThreadPool with just 1 thread will lead to a performance overhead compared with just running the Tasks in the current thread?

evansiroky replied:
Yes, it appears that the ThreadPool does have slightly worse performance with the HSL setup. Over 10 runs, the code that didn't use a ThreadPool was on average 9 seconds faster. I have added this code to this PR.

evansiroky commented:
@optionsome, see my responses to your comments

> I had time now to test this locally. With HSL area data (https://karttapalvelu.storage.hsldev.com/hsl.osm/hsl.osm.pbf, https://elevdata.blob.core.windows.net/elevation/hsl/hsl-10m-elevation-model_20190920.tif and set "elevationUnitMultiplier": 0.1 in build-config.json) 

Thank you for the data.

> it took about double the time to process the elevation data with these changes which is probably not intended. 

Definitely not intended. I don't know why, but with the HSL configuration in particular, the processing times for the elevation module increase with each additional processor added. The good news is that it's already relatively fast for your particular configuration to use just 1 processor. In light of this, I have refactored my PR in 2 ways.

1. I have added a customizable build config parameter of `elevationModuleParallelism` that by default is set to 1.
2. I have created code paths that will not create any additional threads when processing with just 1 processor.

> I'm not sure if it is due to elevation being set to more edges because before the counter of changed edges ("set elevation on 1150000/1379373 edges") didn't ever reach the total number of edges.

This was a bug that has now been fixed. The total previous count was of all edges in the graph, regardless of whether they were StreetWithElevation edges.

> On the upside, this seems to fix #2792 or at least the build does not crash (used https://elevdata.blob.core.windows.net/elevation/hsl/hsl-10m-elevation-model_20190606.tif for testing this). I didn't look into what the elevation values actually look like as I focused now on testing build times and the caching process.

I'm not quite sure why that doesn't appear to happen anymore.

t2gran commented:
Thank you for the update @ekorsky. Just to let you know, we (@abyrd, @optionsome and me) looked at this yesterday. Andrew was taking notes, so he will post some comments from the discussion.

abyrd commented:
Note: all this should probably by copy-pasted into Javadoc on the relevant classes to save future people the time and effort of figuring all this out.

We reviewed this PR in depth with @t2gran and @optionsome. The core topic was the optimizations added here (parallelizing the assignment of elevation profiles to edges, and persisting the results of those calculations for reuse in later graph builds). Everyone found that the second optimization worked well (though the cache currently has to be manually invalidated when elevation data is updated). So most discussion was about the first optimization, why parallelization was not observed to produce a speedup in some cases, and whether recent commits from Evan enabled the intended speedup.

Part of the confusion comes from the fact that we have two GridCoverageFactory implementations, one for NED elevation data (composed of many tiles), and one using a single GeoTIFF file for the whole region. Many US regions will be using the first, and the Nordic countries are using the second. Both have been updated in this PR but they still work slightly differently.

After @evansiroky 's latest updates, calling GridCoverageFactory#getGridCoverage (on either implementation) now creates a completely new coverage on every call, ensuring that each thread gets its own coverage. This was done because the methods to read data from the coverage are synchronized, though according to code comments the reason was not fully understood when those commits were added.

The grid coverages being produced are of class Interpolator2D, which is a subclass of GridCoverage2D that wraps a grid coverage (2D raster) and interpolates it so values change smoothly between the gridded data points. The methods that read values from an Interpolator2D are synchronized because it uses a single interpolator function instance which may have state. However the Interpolator2D is just a wrapper, and these methods are not synchronized on the underlying coverage (which may well be readable by multiple threads at once).

Making a completely new GridCoverage per thread will fix the synchronization problem but it's likely still very inefficient as every thread probably loads its own copy of the elevation raster data into memory. Having N copies of the coverage will limit the benefits of parallelization. With 8 cores memory consumption would be multiplied by 8, and these would be 8 copies of exactly the same data. So there are opposing forces here: on one hand threads can calculate concurrently without lock contention, but on the other hand they're now competing to pull N different identical copies of read-only data into processor cache (though this may not be the case if every coverage is memory-mapping the same underlying disk files, in which case if we're lucky the OS might alias all N in-memory copies to the same physical disk cache pages). We might even be exhausting the heap and increasing garbage collection overhead as each new thread is added.

But I don't think we actually need N copies of the underlying raster. We should be able to use N independent, thread-local thin interpolator wrappers around the same shared underlying raster.

The **GridCoverageFactory for single GeoTIFFs** makes a grid coverage wrapping the GeoTIFF, then wraps that in an Interpolator2D coverage with a bilinear interpolator instance. It should be entirely possible to instead make one GeoTIFF coverage, store it in a field, then make N Interpolator2D coverages (each with its own possibly stateful, synchronized interpolator instance) all wrapping that same GeoTIFF coverage data source.

On the other hand, the **GridCoverageFactory for NED tiled elevation data** is more complex, in that it does two levels of wrapping instead of one. Each tile (a non-interpolated coverage seemingly readable by multiple threads concurrently) is wrapped in its own interpolator coverage (synchronized, single-threaded), and then all those coverages are stitched together into one big coverage with OTP's custom UnifiedGridCoverage. I would not expect this three-level conglomerate to experience significant lock contention as long as all the threads were hitting different tiles, but to the degree that all the threads are calculating elevation profiles for edges in the same geographic area there will be contention. Although the method currently used to find the tile for a given geographic point requires many tiles to be examined, those checks are performed using only unsynchronized methods retrieving the tiles' envelopes, without spurious locking. If we moved the interpolation up one level, outside the UnifiedGridCoverage, each thread could make its own interpolator and they should all be able to operate concurrently with no contention.

Besides eliminating contention for locked interpolators (by making one interpolator per thread), there is another reason to swap the interpolator outside the tile-unifier in the stack of coverage wrappers: each individual tile interpolator currently cannot see the tiles adjacent to it, so I suspect interpolation fails on the edges of each tile. There would be either missing values or constant values outside the edge pixels in each tile, probably leading to jump discontinuities in the unified coverage. Placing the interpolator wrapper outermost should allow it to interpolate across tile boundaries.

There is another issue here in the UnifiedGridCoverage that I hinted at above: finding the correct tile from which to read the data uses a linear scan through a list, which theoretically scales poorly. The more NED tiles are in use, the more time it will spend looking for the right tile; when reading data points from the last tile in the list, it will try and fail on every other tile before finally fetching a value from the last tile, and this is repeated at every single point evaluated. Fortunately this search is no longer being done with repeated exceptions (which are extremely slow), but it's still looping through a lot of failing conditionals. This could be replaced with a constant-time spatial index, but before introducing this complexity we'd first want to measure that all these failed lookups were actually a major source of delay. I'm not sure how many NED tiles we're typically using, and for short lists, linear searches can be faster than hash or tree lookups. They just scale badly to very long lists.

The main conclusion is: for both implementations of the grid coverage factory, ideally only the following method should be called per-thread, because only the interpolator function requires synchronization, while the functions to read the wrapped coverage do not:
coverage = Interpolator2D.create(coverage, new InterpolationBilinear());

I think all of this would also be simpler to follow if we could use parallel streams, letting the JVM manage the thread pool, rather than making our own thread pool. The reason we currently can't use parallel streams is that we need a custom worker thread to hold the thread-specific coverage instance. There should be some other way to do this (e.g. using thread local variables).

evansiroky commented:
@abyrd, @t2gran  and @optionsome thanks for the thorough review. Here are my responses below:

Regarding the idea to reuse the GridCoverage2D instance and instead only create new Interpolator2D instances per thread, I have implemented most of your idea for this change. In the `NEDGridCoverageFactoryImpl` class, I went back to the original implementation of creating a single reusable `UnifiedGridCoverage` instance comprised of multiple `Interplator2D` instances that wrap the `GridCoverage2D` produced by the `GeotiffGridCoverageFactoryImpl` class. Then, in the `ElevationModule` class, a new `Interpolator2D` instance gets created for each thread when the grid coverage factory is the `GeotiffGridCoverageFactoryImpl` class. This idea did end up making things slightly faster despite there being slightly more locking in threads.

However, I am not sure how I would go about implementing your suggestion of creating an `Interpolator2D` instance for a `UnifiedGridCoverage`. The `Interpolator2D` class seems to require a single `GridCoverage2D` instance to get created. I don't see a way to create an `Interpolator2D` instance using multiple `GridCoverage2D` instances. If you can clarify how that might be done or show an example, that would be appreciated.

I refactored the `DegreeGridNEDTileSource` so that it will fail hard except when a tile is missing in the s3 bucket.

I agree that it would take quite a bit of effort to try to automatically calculate the validity of the cached elevation data and that most users probably don't update their underlying elevation data and settings often enough to require invalidating the elevation data. However, I have added some more comments and documentation to note this possibility.

gmellemstrand commented:
I have tested the newest commit with the same dataset as before. I do see some more thread blocking in YourKit compared to `b1f4313`, but total graph building time is now down to 3.3 mins.

I also checked the peak memory consumption and compared to `dev-1.x` it has increased from 3.8GB to 6GB.

This is using `elevationModuleParallelism` of 8 on a 2018 Macbook Pro.

abyrd commented:
Just recording some notes from today's discussion, which we can go over on our next call with @evansiroky.

When you say "This idea did end up making things slightly faster despite there being slightly more locking in threads", was it with the NED or single GeoTIFF data source that you saw this speedup? Actually the goal of my proposed change was not speedup (which was guaranteed) but equal or slightly better performance with N times lower memory usage (i.e. maintaining the memory usage of a single thread). With the tiled NED data it would be expected to cause more lock contention (depending on the number of tiles and order in which the threads evaluate points) but use much less memory and less memory bus saturation. This is a trade-off, and if your speedup was observed on NED data this is encouraging because we are probably getting faster calculations with much less memory consumption.

On the other hand, with the GeoTIFF data source I would expect the lock contention to be the same as before the change (zero lock contention: each thread had and continues to have its own interpolator). But again the memory consumption is expected to be much lower, which reduces memory bus contention, garbage collection pressure, etc. I can't readily see why the amount of waiting for locks would increase, or indeed why any waiting would occur.

@evansiroky and @gmellemstrand, when you refer to the amount of locking, are you referring to the amount of locking itself (amount of time spent inside a synchronized block), or the amount of lock contention (amount of time threads spend waiting on another thread to release a lock)? The first one has near zero cost, and only the second is cause for concern. 

> Then, in the ElevationModule class, a new Interpolator2D instance gets created for each thread when the grid coverage factory is the GeotiffGridCoverageFactoryImpl class.

I know this is just preserving behavior in the original code, but I find it strange to have this `instanceof` check and interpolator wrapping occurring in the ElevationModule. We already have an interface for GridCoverageFactories, I think that interface's contract should just be updated to say it must produce a GridCoverage already optimized for access by a single thread in a pool, and push all Interpolator creation down into the factory implementations. This would be a cleaner use of polymorphism; in the current case we're doing partial polymorphism combined with a more hackish runtime type check in the module using the coverage factories.

> However, I am not sure how I would go about implementing your suggestion of creating an Interpolator2D instance for a UnifiedGridCoverage. 

I'm not sure how you'd do this either. Interpolator2D is supplied by the library and must wrap a GridCoverage2D. Only UnifiedGridCoverage is a class we created and under our control. The interpolation logic in Interpolator2D is too complex to be replicated. We may be reaching the break-even point where the speed advantage is not worth the additional cost of additional development, complexity, and maintenance overhead. The only problem I see is that I don't know how the current implementation would interpolate correctly at tile borders - I suspect it doesn't. 

The trade-off is pretty different for NED vs. single-geotiff data sources. I think we can achieve near-maximum multithreading throughput for a single Geotiff with the latest commits, but NED data would take more work.

There may be a relatively simple way to implement this for NED data though: the first thing I'd try would be to make a UnifiedGridCoverage for each thread, have each of those UnifiedGridCoverages hold its own Interpolator2D for each tile, but share the underlying GridCoverage2D for each tile across all threads. So your coverage factory would need to build up a list of GridCoverage2Ds only once, then build up a separate set of Interpolator2Ds and a containing UnifiedGridCoverages each time it was asked to produce a coverage for a new thread.

On further inspection though, I'm not entirely sure whether each Interpolator2D is storing large amounts of data (on the order of the entire Coverage2D it's wrapping) or not. That needs to be understood to really judge the effectiveness of my proposed changes from a memory consumption point of view.

A few final points that might be clarified more quickly by discussion:
- There are references to "missing" elevation data in the code - which elevation data is not missing before we read from our elevation source? Is the non-missing data just elevation tags in OSM?
- Why is the interpolator lazy-initialized instead of being created immediately upfront? Why does that lazy initialization process require several layers of nested method calls?
- Does the most recent commit af4130c "Add EllipsoidToGeoidDifference for known elevations in OSM" add new functionality or restore functionality that was somehow lost in the other changes?

If I have time I hope to make a branch with some suggested implementations before our call next week. Hope to talk to you then!

gmellemstrand commented:
> @evansiroky and @gmellemstrand, when you refer to the amount of locking, are you referring to the amount of locking itself (amount of time spent inside a synchronized block), or the amount of lock contention (amount of time threads spend waiting on another thread to release a lock)? The first one has near zero cost, and only the second is cause for concern.
> 

YourKit says state: blocked. I couldn't find any documentation on that specific status message, but I assume that means it is waiting for a lock to be released. There is not a lot of time being spent waiting, but it is more than before.

abyrd commented:
> YourKit says state: blocked. I couldn't find any documentation on that specific status message, but I assume that means it is waiting for a lock to be released. There is not a lot of time being spent waiting, but it is more than before.

We just looked at the profiling report together. The blocking is very infrequent, sometimes over a minute passes between threads blocking each other. The blocking occurs deep in the Java imaging libraries where they are loading tiles of the underlying image (not our NED tile implementation, but tiles of one single GeoTIFF image). This makes sense as a side effect of the most recent changes: all threads are now sharing a single underlying image, which seems to be lazy-loaded in pieces, and the loading code seems to block all reads of image data while it's loading pieces of the image.

Previously when each thread had its own underlying image, they all idled separately loading and decoding their copies of the image and did not block each other. Now with the shared data and blocking, the total time the elevation calculations are idle while loading the image should be the same or even less than previously, since the image is decoded only once and uses N times less memory. So this is a natural and even desirable sign that the threads are sharing data.

So with the shared source image and one interpolator per thread, total graph building time is now down to 3.3 mins; with a separate source image per thread time was 5.9 minutes; with a single interpolator shared across all threads it was 11.4 minutes. So we have reduced it over 70%. Not bad, I think we can consider the single-geotiff multithreading implementation complete.

evansiroky commented:
@abyrd thanks for the additional clarification and suggestions related to the UnifiedGridCoverage. I have gone ahead and implemented these changes and noticed a sizeable speed-up with the TriMet graph that uses this configuration. I don't know much about the interpolation at the edges to be able to comment one way or another, but that issue seems kind of out-of-scope for this PR since the underlying calculations of elevation are using the same data and methods.

Also, see my responses to some of your comments below:

> I think all of this would also be simpler to follow if we could use parallel streams, letting the JVM manage the thread pool, rather than making our own thread pool. The reason we currently can't use parallel streams is that we need a custom worker thread to hold the thread-specific coverage instance. There should be some other way to do this (e.g. using thread local variables).

Oh wow, I didn't know what ThreadLocal was until you mentioned it. I also went ahead and refactored my code to use Thread Local variables instead of the fork/join pool stuff and things look much cleaner now. However, I had to make an all-or-just-one-thread parameter instead of allowing different numbers of threads to be configured. This should be fine because the whole reason this parameter exists anyways is for the Helsinki graph build which is always best with just one thread anyways. All other graph builds seem to work best with the maximum amount of threads available.

> There are references to "missing" elevation data in the code - which elevation data is not missing before we read from our elevation source? Is the non-missing data just elevation tags in OSM?

Yes, the non-missing data is from tags in OSM. They are found and entered here:

https://github.com/opentripplanner/OpenTripPlanner/blob/ca79a24c8b2f1e53368592b0bb6928532d440553/src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java#L636-L658

...and then added to the extra graph build module data here:

https://github.com/opentripplanner/OpenTripPlanner/blob/ca79a24c8b2f1e53368592b0bb6928532d440553/src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java#L258

> Why is the interpolator lazy-initialized instead of being created immediately upfront? Why does that lazy initialization process require several layers of nested method calls?

Creating each of the interpolators immediately upfront blocks other threads as the creation of an interpolator must be synchronized. Because of this blocking, there are 2 ways this affects performance when using cached elevation data from a previous graph build. 

1. The interpolators may not even be needed if the cached elevations file contains data for all edges.
2. By lazy-creating the interpolators, some threads that encounter lots of data for which cached data is available will be able to immediately proceed with setting data instead of waiting for a lock to be released at the very beginning.

> Does the most recent commit af4130c "Add EllipsoidToGeoidDifference for known elevations in OSM" add new functionality or restore functionality that was somehow lost in the other changes?

This is part of a somewhat larger feature that adds the ability to include the EllipsoidToGeoidDifference in the calculations of elevations on edges. Prior to this PR, adding this value was only possible when making requests, and that value was only available as a single graph-wide value. After planning a few trips with a graph built with this data, I noticed that the EllipsoidToGeoidDifference wasn't also being added to elevation data from OpenStreetMap, so I made this change.

gmellemstrand commented:
Another thing that I think we could change in this pull request. Should this log message be info instead of error?

https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/src/main/java/org/opentripplanner/routing/util/ElevationUtils.java#L74
